Implement equality and hashing for unions

icedtrees opened the pull request:
Unions should be considered equal to each other if they subclass each other and have the same tag and value. See proposal at https://paper.dropbox.com/doc/APIv2-Union-and-namedtuple-2MZPn0EpJVfYKw5pa4boo?t=549589594649114#:t=549589594649114

cakoose commented on stone/target/python_rsrc/stone_base.py:
```diff
@@ -34,6 +34,16 @@ def __init__(self, tag, value=None):
+        # Also need to check if one class is a subclass of another. If one union extends another,
+        # the common fields should be able to be compared to each other.
+        return (
+            (isinstance(self, other.__class__) or isinstance(other, self.__class__)) and
```
Why not just check that the classes are the same?  Do we ever want a value to be "equal" to a value of a subclass?

icedtrees replied:
Yes. See discussion at https://tails.corp.dropbox.com/D227666:

union A
   t0

union B extends A
   t1

A(t0) should be equal to B(t0).

cakoose replied:
Oh, that makes sense.  Though now I'm a bit worried about "identical" values having different underlying classes.

When someone instantiates says `B.t0`, maybe we should return `A.t0`?

braincore replied:
In the simplest case of referencing the auto-populated `B.t0`, it is already a reference to `A.t0`. But, it's possible to manually instantiate `B` to `t0`, and it would take some generator work to reject that. I think out of scope for this diff. Also, the deserializer uses the manual instantiation method :( so that'll have to change.

cakoose commented on stone/target/python_rsrc/stone_base.py:
```diff
@@ -34,6 +34,16 @@ def __init__(self, tag, value=None):
+
+    def __hash__(self):
+        return hash((self._tag, self._value))
 
```
The Python docs say to implement `__ne__` if you implement `__eq__`.  (It's fine to implement `__ne__` by calling `__eq__`.)

cakoose commented on stone/target/python_rsrc/stone_base.py:
```diff
@@ -34,6 +34,16 @@ def __init__(self, tag, value=None):
         self._tag = tag
         self._value = value
 
+    def __eq__(self, other):
```
If someone passes in an instance of `object()`, this function will fail with a runtime error, right?  I'm not sure what the contract of `__eq__` is, but maybe you're required to return `false` in that case?

braincore replied:
You should return false if `other` isn't a subclass of `Union`.

braincore commented:
Sorry for the delay, was on PTO for a couple of weeks. Once you address @cakoose's comments, this is good to go.

icedtrees commented:
Sorry for late follow-up! Addressed comments. Thanks for thorough review!

I'm not part of the Dropbox github team, so you'll have to add me or merge the PR.

braincore commented:
Can you please squash your commits? I'm going to try this new "Rebase and Merge" GitHub feature :)

icedtrees commented:
Done!
